data/preprocessor.js

```diff
@@ -0,0 +1,98 @@
+/* eslint-disable no-use-before-define */
+
+const Preprocessor = (jsonRaw = {}) => {
+  try {
+    const canRenderTimestamp = checkInputContainsProperties({
+      input: jsonRaw,
+      properties: [
+        'metadata.firstPublished',
+        'metadata.lastUpdated',
+        'content.model.blocks',
+      ],
+    });
+    if (canRenderTimestamp) {
+      // construct a new block from the metadata
+      const timestampBlock = {
+        type: 'timestamp',
+        model: {
+          published: jsonRaw.metadata.firstPublished,
+          updated: jsonRaw.metadata.lastUpdated,
+        },
+      };
+      return insertTimestampBlock(jsonRaw, timestampBlock);
+    }
+  } catch (e) {
+    // if our block manipulation fails for whatever reason, fall back to jsonRaw
+  }
+  return jsonRaw;
+};
+
+export const checkInputContainsProperties = ({ input, properties }) =>
+  properties.reduce(
+    (validates, prop) => validates && inputContainsProperty(input, prop),
+    true,
+  );
+
+const deepClone = originalObj => JSON.parse(JSON.stringify(originalObj));
+
+/**
+ * Checks if Object has property by name.
+ * Properties can be single ('foo.bar') or chained ('foo.bar.baz')
+ * So can run `if (inputContainsProperty(myInput, 'foo.bar.baz))` to check for child property.
+ * Saves having to do `if (foo && foo.bar && foo.bar.baz) `.
+ * @param {Object} input e.g. { foo: { bar: 'baz' } }
+ * @param {String} prop e.g. 'foo.bar'
+ */
+const inputContainsProperty = (input, prop) => {
```

Comment:
Im a big fan of 
```
const get = (path, object) => path.reduce((xs, x) => xs && xs[x] ? xs[x] : null, object);
```
e.g.
```
const  firstPublished = get(['metadata', 'firstPublished'], jsonRaw)
```

Source: https://medium.com/javascript-inside/safely-accessing-deeply-nested-values-in-javascript-99bf72a0855a

Reply:
I'm less keen on that as it looks like we're grabbing two top-level properties called `metadata` and `firstPublished` 😅 

But I'm also not keen on having to write this `inputContainsProperty` function in Simorgh. Almost leaning towards ripping it out and just doing `if (metadata && metadata.firstPublished && ...` as it's one-time logic we're using specifically for the timestamp placement. It'll look gross but it'll be easy to mentally parse.

Currently 40:40:20 on a) keeping as-is, b) stripping it out, c) adopting your `get` solution 🤔

Reply:
Second thoughts, coming round to the idea 😄 will just need some careful commenting. Less code = better in this case.